main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -180,8 +180,67 @@ object Safety {
     case Expression.Cast(exp, _, _, _, _, _, _, _) =>
       visitExp(exp)
 
-    case Expression.Upcast(exp, tpe, pur, eff, loc) =>
-      throw InternalCompilerException("Not implemented")
+    case e@Expression.Upcast(exp, tpe, pur, eff, loc) =>
```

Comment:
This looks like a lot of fun :) :)

But try write a recursive function on that matches on `expected` and `actual` simultaneously. 

You will find this a lot easier, I think.

Reply:
Like this:

```
def isSafeUpcast(expected: Type, actual: Type): Bool = (expected, actual) match {
  case (Type.Cst(...), Type.Cst(...)) => // do checks
  case (Type.Apply(...), Type.Apply(...) => // recurse on the parts
/// and so forth
```

Reply:
So you line them up one at a time